enhancements/builds/node-cache-for-builds.md

```diff
@@ -0,0 +1,186 @@
+---
+title: node-cache-for-builds
+authors:
+  - "@adambkaplan"
+reviewers:
+  - "@nalind
+  - "@rhdan"
+  - "@TomSweeneyRedHat"
+  - "@bparees"
+approvers:
+  - "@bparees"
+creation-date: 2019-02-18
+last-updated: 2019-02-18
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Node Cache for Builds
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+None.
+
+## Summary
+
+Builds in OpenShift 4.x currently require every layer of the base image to be pulled from its source
+registry every time the build is run. In OCP 3.x builds could take advantage of image layers that
+were stored on the build's node. This proposal provides a means of partially restoring this
+node-level cache by allowing build pods to avoid pulling layers that are present on the node. Unlike
+3.x, however, layers created during the build process will not be directly cached on the node due to
+buildah's isolation.
+
+## Motivation
+
+### Goals
+
+1. Provide a means for builds to take advantage of image layers that are stored on the node.
+
+### Non-Goals
+
+1. Provide a means for builds to write their layers to a cache (node or shared).
+2. Provide a means for builds to share an image layer cache within a namespace.
+3. Provide a means for the internal registry to pre-populate an image layer cache within a namespace
+   or across a cluster.
+4. Add persistent volume support for builds.
+
+## Proposal
+
+Build pods will mount the image layer cache that exists on every OpenShift node into the build pod.
+Buildah - running within the build pod - will use this layer cache to reduce the number of layers
+that need to be pulled when the build commences.
+
+### User Story
+
+As a developer using OpenShift to build container images I want builds to use image layers cached on
+the node So that builds don't have to pull all layers in their base images And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at `/var/lib/containers/storage`
```

Comment:
I have a question about this: If I assume that the images used during a build process  are not images the cluster generally runs as pod when are we going to have a hit on the cache? 

e.g. if I use a golang image during a build this image wont be cached(as the cache is readonly) and as i never use to run this very same golang image in a Deployment/Pod it will never be cached.

Reply:
well if you run an image that's based on that golang image, the base layer would still get into the cache.

but yes if you use it as a base for a multistage build you're correct.  or if you just build the image and push it but never deploy it.  (adam has some references to these scenarios elsewhere in the doc)

that said, in many cases the the build is based on "ubi" or "centos" in which case we would still get benefit.

Reply:
(or even a ruby or nodejs base image, which when the app image that's built on top of it gets deployed, would result in the ruby/nodejs base getting pulled into the cache)

Reply:
@ricardomaraschini I call this out in the drawbacks section. S2I builds will likely benefit the most because these tend to produce "fat" images that are loaded with test/dev tools. Multistage Dockerfile builds will likely benefit the least, since developers often use these to produce "skinny"/"runtime" images.